v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/compiler/index.ts

```diff
@@ -79,8 +86,71 @@ export async function downloadConfiguredCompilers(
 
 export async function getCompiler(
   version: string,
-  { preferWasm }: { preferWasm: boolean },
+  { preferWasm, compilerPath }: { preferWasm: boolean; compilerPath?: string },
+): Promise<Compiler> {
+  if (compilerPath !== undefined) {
+    // If a compiler path is provided, it means the user is using a custom compiler
+    return getCompilerFromPath(version, compilerPath);
+  } else {
+    // Otherwise we get or download the compiler for the specific version
+    return getCompilerFromVersion(version, { preferWasm });
+  }
+}
+
+async function getCompilerFromPath(
+  compilerVersion: string,
+  compilerPath: string,
 ): Promise<Compiler> {
+  log(`Using custom compiler ${compilerPath}`);
+  if (!(await exists(compilerPath))) {
+    throw new HardhatError(
+      HardhatError.ERRORS.CORE.SOLIDITY.COMPILER_PATH_DOES_NOT_EXIST,
+      { compilerPath, version: compilerVersion },
+    );
+  }
+
+  const isWasm = !(await isBinaryFile(compilerPath));
+
+  log(`Using ${isWasm ? "WASM" : "Native"} compiler`);
+
+  const execFileAsync = promisify(execFile);
+
+  let stdout: string;
+
+  if (isWasm) {
+    const solc = (await import(pathToFileURL(compilerPath).toString())).default;
+    const { version } = wrapper(solc);
+    stdout = version();
+  } else {
+    stdout = (await execFileAsync(compilerPath, ["--version"])).stdout;
+  }
+
+  log(`Version output: ${stdout}`);
+
+  const match = stdout.match(/(?<longVersion>\d+\.\d+\.\d+\+commit\.\w+)/);
+
+  if (match === null || match.groups === undefined) {
+    throw new HardhatError(
+      HardhatError.ERRORS.CORE.SOLIDITY.PARSING_VERSION_STRING_FAILED,
+      { versionString: stdout },
+    );
+  }
+
+  const { longVersion } = match.groups;
```

Comment:
I'm not convinced about this. Some questions:

- How sure are we that the output of solc's `--version` won't change?
- Does this work with solx?
- Do we actually need this? What happens if we just use `compilerVersion` as the long version in the `SolcJsCompiler`/`NativeCompiler` constructors? Would that be more or less risky than relying on the `--version` output?

Reply:
- The --version output format has been always the same since the first version I could check (0.4.11). It could change but seems unlikely
- It works with solx and solcjs
- We'd need to ask pato on the implications on not having the proper long version

Reply:
I'm ok with keeping this approach if necessary (I agree with Armando that it seems unlikely to change, and if it does it will fail in a easy to detect way). But right now I don't think I can make a call on the trade-off because I don't know how important it is to have the proper long version.